core/commands/unixfs/ls.go

```diff
@@ -0,0 +1,199 @@
+
+For files, the child size is the total size of the file contents.  For
+directories, the child size is the IPFS link size.
+`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("ipfs-path", true, true, "The path to the IPFS object(s) to list links from").EnableStdin(),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.Context().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		paths := req.Arguments()
+
+		output := LsOutput{
+			Arguments: map[string]string{},
+			Objects:   map[string]*LsObject{},
+		}
+
+		for _, fpath := range paths {
+			ctx := req.Context().Context
+			merkleNode, err := core.Resolve(ctx, node, path.Path(fpath))
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			key, err := merkleNode.Key()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			hash := key.B58String()
+			output.Arguments[fpath] = hash
+
+			if _, ok := output.Objects[hash]; ok {
+				// duplicate argument for an already-listed node
+				continue
+			}
+
+			output.Objects[hash] = &LsObject{}
+
+			unixFSNode, err := unixfs.FromBytes(merkleNode.Data)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			t := unixFSNode.GetType()
+			switch t {
+			default:
+				res.SetError(fmt.Errorf("unrecognized type: %s", t), cmds.ErrImplementation)
+				return
+			case unixfspb.Data_File:
+				key, err := merkleNode.Key()
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+				output.Objects[hash].Links = []LsLink{LsLink{
+					Name: fpath,
+					Hash: key.String(),
+					Type: t.String(),
+					Size: unixFSNode.GetFilesize(),
+				}}
+			case unixfspb.Data_Directory:
+				links := make([]LsLink, len(merkleNode.Links))
+				output.Objects[hash].Links = links
+				for i, link := range merkleNode.Links {
+					getCtx, cancel := context.WithTimeout(ctx, time.Minute)
+					defer cancel()
+					link.Node, err = link.GetNode(getCtx, node.DAG)
+					if err != nil {
+						res.SetError(err, cmds.ErrNormal)
+						return
+					}
+					d, err := unixfs.FromBytes(link.Node.Data)
+					if err != nil {
+						res.SetError(err, cmds.ErrNormal)
+						return
+					}
+					t := d.GetType()
+					lsLink := LsLink{
+						Name: link.Name,
+						Hash: link.Hash.B58String(),
+						Type: t.String(),
+					}
+					if t == unixfspb.Data_File {
+						lsLink.Size = d.GetFilesize()
+					} else {
+						lsLink.Size = link.Size
+					}
+					links[i] = lsLink
+				}
+			}
+		}
+
+		res.SetOutput(&output)
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+
+			output := res.Output().(*LsOutput)
+			buf := new(bytes.Buffer)
+			w := tabwriter.NewWriter(buf, 1, 2, 1, ' ', 0)
+
+			nonDirectories := []string{}
+			directories := []string{}
+			for argument := range output.Arguments {
+				hash := output.Arguments[argument]
+				object := output.Objects[hash]
```

Comment:
On Tue, Jun 16, 2015 at 04:31:31PM -0700, Juan Batiz-Benet wrote:
> 2. the CLI may be used against any process implementing the IPFS
>    API. there is zero guarantee it is the same code that is in
>    go-ipfs.

Where does that split land?  I'd expect both the Run and Marshalers
handling to happen on the daemon side, so they need to be in sync with
each other, but not necessarily with any client code.

I'll reroll to check the map-access (since I understand points (1) and
(3)), but if there's some important kernel of wisdom in point (2), I'd
like to clear up my confusion there ;).